### PR TITLE
Enable vectorcall on limited api cyfunctions

### PR DESCRIPTION
**Pull request by da-woods**
Python 3.12 and upwards (since this is where the limited API exposes the protocol). In principle we could probably use the "backport vectorcall" mechanism, but that doesn't seem in the spirit of the limited api.

### CODE REVIEW

File: Cython/Utility/CythonFunction.c [hunk 1]
```diff
@@ -787,36 +792,24 @@ static PyObject * __Pyx_CyFunction_CallMethod(PyObject *func, PyObject *self, Py
...
-    Py_DECREF(py_name);
-#else
-    PyErr_Format(PyExc_TypeError, "%.200s() takes no keyword arguments",
-                 f->m_ml->ml_name);
-#endif
+    name = __Pyx_CyFunction_get_name_for_format(func);
+    if (unlikely(!name)) return NULL;
+    PyErr_Format(PyExc_TypeError, __Pyx_FORMAT_NAME "() takes no keyword arguments",
+                 name);
+    __Pyx_DECREF_name_for_format(name);
```
Comment (scoder) [inline Cython/Utility/CythonFunction.c:812]: This exception generation code is used a lot. Seems worth moving it into a helper function. Or two, given the second case with the argument size. That would also centralise the "get name + decref" code and avoid additional global declarations for it.

(I considered [PyErr_FormatV](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_FormatV), but I don't think that fits nicely with additionally passing the function name.)

**Comment by scoder** [comment]
This is great to have, it might really make a visible difference in performance.

File: Cython/Utility/CythonFunction.c (hunk 1 above)
Comment (da-woods) [reply Cython/Utility/CythonFunction.c]: I've turned it into a couple of helper macros. I was struggling to make the "size" argument version work with a function. Hopefully that's a bit clearer.

Comment (scoder) [reply Cython/Utility/CythonFunction.c]: I was thinking of something like this:
```c
static void __Pyx_CyFunction_raise_argument_count_error(PyObject *func, const char* message, Py_ssize_t size) {
    PyObject *py_name = __Pyx_CyFunction_get_name((__pyx_CyFunctionObject*)func, NULL);
    if (!py_name) return;
    PyErr_Format(PyExc_TypeError,
        "%.200S() %.50s (%" CYTHON_FORMAT_SSIZE_T "d given)",
        py_name, message, size);
    Py_DECREF(py_name);
}

static void __Pyx_CyFunction_raise_type_error(PyObject *func, const char* message) {
    PyObject *py_name = __Pyx_CyFunction_get_name((__pyx_CyFunctionObject*)func, NULL);
    if (!py_name) return;
    PyErr_Format(PyExc_TypeError,
        "%.200S() %.50s",
        py_name, message);
    Py_DECREF(py_name);
}
``

**Review by scoder — APPROVED** [review]

**Pull request merged**